File: arrays_and_strings/48_rotate_image/Python/rotate_image.py
from typing import List
import copy as c

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)-1
        copy = c.deepcopy(matrix)
        
        for index, _ in enumerate(copy):
            new_row = []
            for i in range(n, -1, -1):
                new_row.append(copy[i][index])
            matrix[index] = new_row

if __name__ == "__main__":
    test_cases = [
        [[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]],

        [[5, 1, 9, 11],
         [2, 4, 8, 10],
         [13, 3, 6, 7],
         [15, 14, 12, 16]],

        [[1]],

        [[1, 2],
         [3, 4]],
    ]

    for matrix in test_cases:
        import copy
        original = copy.deepcopy(matrix)
        Solution().rotate(matrix)
        print(f"Input:       {original}")
        print(f"Output:      {matrix}")
        print("-" * 35)

File: arrays_and_strings/48_rotate_image/TS/rotateImage.ts
export {};

function rotate(matrix: number[][]): void {
  // if(matrix.length !== matrix[0].length) return;
  const n = matrix.length - 1;

  let lastRow: number[] = [];
  const copy = matrix.map(r => [...r]);

  copy.forEach((_, r) => {
    let newRow: number[] = [];
    for (let i = n; i >= 0; i--) {
        newRow.push(copy[i][r]);
    }
    matrix[r] = newRow;
  });
}

const testCases: number[][][] = [
  [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
  ],

  [
    [5, 1, 9, 11],
    [2, 4, 8, 10],
    [13, 3, 6, 7],
    [15, 14, 12, 16],
  ],

  [[1]],

  [
    [1, 2],
    [3, 4],
  ],
];

for (const matrix of testCases) {
  const original = matrix.map((row) => [...row]);
  rotate(matrix);
  console.log(`Input:       ${JSON.stringify(original)}`);
  console.log(`Output:      ${JSON.stringify(matrix)}`);
  console.log("-".repeat(35));
}